tsdb/engine/tsm1/engine.go

```diff
@@ -939,106 +911,164 @@ func (e *Engine) compactTSMFull() {
-							return
-						} else if err != nil {
-							e.logger.Printf("error compacting TSM files: %v", err)
-							atomic.AddInt64(&e.stats.TSMFullCompactionErrors, 1)
-
-							time.Sleep(time.Second)
-							return
-						}
-					}
-
-					if err := e.FileStore.Replace(group, files); err != nil {
-						e.logger.Printf("error replacing new TSM files: %v", err)
-						atomic.AddInt64(&e.stats.TSMFullCompactionErrors, 1)
-						time.Sleep(time.Second)
-						return
-					}
-
-					for i, f := range files {
-						e.logger.Printf("compacted %s group (%d) into %s (#%d)", logDesc, groupNum, f, i)
-					}
-
-					if optimize {
-						atomic.AddInt64(&e.stats.TSMOptimizeCompactions, 1)
-					} else {
-						atomic.AddInt64(&e.stats.TSMFullCompactions, 1)
-					}
-					e.logger.Printf("compacted %s %d files into %d files in %s",
-						logDesc, len(group), len(files), time.Since(start))
-				}(i, group)
-			}
-			wg.Wait()
+			s.Apply()
+		}
+	}
+}
 
-			// Track the amount of time spent compacting the groups.
-			if optimize {
-				atomic.AddInt64(&e.stats.TSMOptimizeCompactionDuration, time.Since(start).Nanoseconds())
-			} else {
-				atomic.AddInt64(&e.stats.TSMFullCompactionDuration, time.Since(start).Nanoseconds())
-			}
+// compactionStrategy holds the details of what to do in a compaction.
+type compactionStrategy struct {
+	compactionGroups []CompactionGroup
+
+	fast        bool
+	description string
+
+	durationStat *int64
+	activeStat   *int64
+	successStat  *int64
+	errorStat    *int64
+
+	logger    *log.Logger
+	compactor *Compactor
+	fileStore *FileStore
+}
+
+// Apply concurrently compacts all the groups in a compaction strategy.
+func (s *compactionStrategy) Apply() {
+	start := time.Now()
+
+	var wg sync.WaitGroup
+	for i := range s.compactionGroups {
+		wg.Add(1)
+		go func(groupNum int) {
+			defer wg.Done()
+			s.compactGroup(groupNum)
+		}(i)
+	}
+	wg.Wait()
+
+	atomic.AddInt64(s.durationStat, time.Since(start).Nanoseconds())
+}
+
+// compactGroup executes the compaction strategy against a single CompactionGroup.
+func (s *compactionStrategy) compactGroup(groupNum int) {
+	group := s.compactionGroups[groupNum]
+	start := time.Now()
+	s.logger.Printf("beginning %s compaction of group %d, %d TSM files", s.description, groupNum, len(group))
+	for i, f := range group {
+		s.logger.Printf("compacting %s group (%d) %s (#%d)", s.description, groupNum, f, i)
+	}
+
+	files, err := func() ([]string, error) {
+		// Count the compaction as active only while the compaction is actually running.
+		atomic.AddInt64(s.activeStat, 1)
+		defer atomic.AddInt64(s.activeStat, -1)
+
+		if s.fast {
+			return s.compactor.CompactFast(group)
+		} else {
+			return s.compactor.CompactFull(group)
+		}
+	}()
+
+	if err != nil {
+		if err == errCompactionsDisabled || err == errCompactionInProgress {
+			s.logger.Printf("aborted %s compaction group (%d). %v", s.description, groupNum, err)
 
+			if err == errCompactionInProgress {
+				time.Sleep(time.Second)
```

Comment:
@jwilder it seems like every error case has a `time.Sleep` - but `errCompactionsDisabled` seems like the only error case where we don't sleep.

Do you think it would be okay to remove the sleep calls from `compactGroup` and unconditionally sleep inside `compactTSMLevel` and `compactTSMFull`?